lib/Basic/FileInfo.cpp

```diff
@@ -14,22 +14,93 @@
 
 #include <cassert>
 #include <cstring>
+
+#if defined(_WIN32)
+#include <windows.h>
+#include <direct.h>
+#else
 #include <sys/stat.h>
+#endif
 
 using namespace llbuild;
 using namespace llbuild::basic;
 
 bool FileInfo::isDirectory() const {
-  return mode & S_IFDIR;
+  return (mode & S_IFDIR) != 0;
 }
 
+#if defined(_WIN32)
+static void FileTimeToTimestamp(FILETIME fileTime, FileTimestamp *stamp) {
+  // Windows counts file times in 100-nanosecond intervals since January 1, 1601
+  // (UTC)
+  // UNIX counts file times in seconds since January 1, 1970 (UTC)
+  const uint64_t HUNDREDS_OF_NANOSECONDS_PER_SECOND = 10000000;
+  const uint64_t EPOCH_DIFFERENCE =
+    11644473600 * HUNDREDS_OF_NANOSECONDS_PER_SECOND;
+
+  uint64_t totalHundredsOfNanoseconds =
+    (((uint64_t)fileTime.dwHighDateTime << 32) |
+    (uint64_t)fileTime.dwLowDateTime);
+  totalHundredsOfNanoseconds -= EPOCH_DIFFERENCE;
+
+  stamp->seconds =
+    totalHundredsOfNanoseconds / HUNDREDS_OF_NANOSECONDS_PER_SECOND;
+  stamp->nanoseconds =
+    totalHundredsOfNanoseconds % HUNDREDS_OF_NANOSECONDS_PER_SECOND;
+}
+#endif
+
 /// Get the information to represent the state of the given node in the file
 /// system.
 ///
 /// \param info_out [out] On success, the important path information.
 /// \returns True if information on the path was found.
 FileInfo FileInfo::getInfoForPath(const std::string& path, bool asLink) {
   FileInfo result;
+
+#if defined(_WIN32)
+  DWORD attributes;
+  if (asLink) {
+    attributes = FILE_FLAG_OPEN_REPARSE_POINT;
+  }
+  else {
+    attributes = FILE_ATTRIBUTE_NORMAL;
+  }
+
+  HANDLE fileHandle =
+    CreateFileA(path.c_str(), GENERIC_READ,
+      FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+      nullptr, OPEN_EXISTING, attributes, nullptr);
+
+  _BY_HANDLE_FILE_INFORMATION fileInformation;
+  if (!GetFileInformationByHandle(fileHandle, &fileInformation)) {
+    CloseHandle(fileHandle);
+    memset(&result, 0, sizeof(result));
+    assert(result.isMissing());
+    return result;
+  }
+
+  FileTimeToTimestamp(fileInformation.ftLastWriteTime, &result.modTime);
+  result.inode = 0;
+  result.device = _getdrive() - 1;
+
+  auto fileAttributes = fileInformation.dwFileAttributes;
+  auto mode = S_IREAD;
+  if (fileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+    mode |= S_IFDIR;
+  else
+    mode |= S_IFREG;
+
+  if ((fileAttributes & FILE_ATTRIBUTE_READONLY) == 0)
+    mode |= S_IWRITE;
+
+  result.mode = mode;
```

Comment:
It's not obvious, but we also use it in comparisons, because we directly compare the literal bytes of the `FileInfo` inside the build system.